### ISSUE TITLE
Upgrading from 2.0.1.RELEASE to 2.1.1.RELEASE - causes Buffer underflow when restoring the StateMachineContext

### ISSUE THREAD

**Issue by jonseah** [spring-projects/spring-statemachine#721]
Hello, 

In my project we persist the spring-statemachine to our database via the JpaPersistingStateMachineInterceptor. We are in the process of upgrading from 2.0.1.RELEASE to 2.1.1.RELEASE

The issue we are finding is that when we upgrade to 2.1.1.RELEASE, restoring from the db a statemachine created by 2.0.1.RELEASE would result in the following exception

```
Caused by: com.esotericsoftware.kryo.KryoException: Buffer underflow.
	at com.esotericsoftware.kryo.io.Input.require(Input.java:199)
	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:373)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:127)
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:693)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:804)
	at org.springframework.statemachine.kryo.StateMachineContextSerializer.read(StateMachineContextSerializer.java:65)
	at org.springframework.statemachine.kryo.StateMachineContextSerializer.read(StateMachineContextSerializer.java:38)
	at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:709)
	at org.springframework.statemachine.kryo.KryoStateMachineSerialisationService.doDecode(KryoStateMachineSerialisationService.java:45)
	at org.springframework.statemachine.kryo.AbstractKryoStateMachineSerialisationService$3.execute(AbstractKryoStateMachineSerialisationService.java:147)
	at com.esotericsoftware.kryo.pool.KryoPoolQueueImpl.run(KryoPoolQueueImpl.java:58)
	at org.springframework.statemachine.kryo.AbstractKryoStateMachineSerialisationService.decode(AbstractKryoStateMachineSerialisationService.java:143)
	at org.springframework.statemachine.kryo.AbstractKryoStateMachineSerialisationService.decode(AbstractKryoStateMachineSerialisationService.java:130)
	at org.springframework.statemachine.kryo.AbstractKryoStateMachineSerialisationService.deserialiseStateMachineContext(AbstractKryoStateMachineSerialisationService.java:72)
	at org.springframework.statemachine.data.RepositoryStateMachinePersist.read(RepositoryStateMachinePersist.java:76)
	at org.springframework.statemachine.data.jpa.JpaPersistingStateMachineInterceptor.read(JpaPersistingStateMachineInterceptor.java:70)
	at org.springframework.statemachine.persist.AbstractStateMachinePersister.restore(AbstractStateMachinePersister.java:70)
```
The issue boils down to line 65 of the StateMachineContextSerializer. It appears that in a later release childRefs was introduced to the StateMachineContext and thus causing a deserialization issue as 2.0.1.RELEASE does not have this childRef
https://github.com/spring-projects/spring-statemachine/blob/cc0550d0d54887e4ecf58e94b9140a13434b36ad/spring-statemachine-kryo/src/main/java/org/springframework/statemachine/kryo/StateMachineContextSerializer.java#L55-L68

Is there any workaround for this? or perhaps is there some form of migration of our existing StateMachineContexts to 2.1.1.RELEASE?

**Comment by jonseah** [comment]
One way I believe we can resolve this is by overriding the StateMachineContextSerializer.read with the version from 2.0.1.RELEASE and force a write with the 2.1.1.RELEASE version.

Is this a feasible approach?

**Comment by jonseah** [comment]
Another approach is that we could make the latest code of StateMachineContextSerializer backwards compatible with the 2.0.1.RELEASE code

so
```
List<String> childRefs = new ArrayList<>();
if(input.canReadInt()) {
    childRefs = (List<String>) kryo.readClassAndObject(input);
} 
```
Will send a PR over when I get home

**Comment by jonseah** [comment]
whoops, accidently closed

**Comment by hamid646** [comment]
I had this issue too, it’s basically Kryo upgrade issue. If you have data saved using Kryo version 3.X can not be read by Kryo version 4, which is part of new SM version. So i ended up truncating existing data.

**Comment by jonseah** [comment]
I've debugged the code and as per what i mentioned in my earlier comment, it is due to the addition of the the childRefs in spring-statemachine 2.1+

to reproduce here's my test
```
	@Test
	public void testBackwardsCompatibiilty() throws Exception {
	    // given
	    String encodedString = "AQMBZXZlbvQDAXN0YXTlAAEAb3JnLnNwcmluZ2ZyYW1ld29yay5zdGF0ZW1hY2hpbmUuc3VwcG9ydC5PYnNlcnZhYmxlTWHwAQABAWphdmEudXRpbC5BcnJheUxpc/QBAAECamF2YS51dGlsLkhhc2hNYfABAAA=";
		byte[] bytes = Base64.getDecoder().decode(encodedString);
		KryoStateMachineSerialisationService<String, String> service = new KryoStateMachineSerialisationService<>();

		// when
		StateMachineContext<String, String> stateMachineContext = service.deserialiseStateMachineContext(bytes);

		// then
		assertThat(stateMachineContext, is(notNullValue()));
	}
```
The encodedString is a Base64 encoded version of the kyro serialization based on spring-statemachine 2.0.1.RELEASE. When this test is run against spring-statemachine 2.1.1.RELEASE it fails with the expected 

```
com.esotericsoftware.kryo.KryoException: Buffer underflow.
	at com.esotericsoftware.kryo.io.Input.require(Input.java:199)
	at com.esotericsoftware.kryo.io.Input.readVarInt(Input.java:373)
	at com.esotericsoftware.kryo.util.DefaultClassResolver.readClass(DefaultClassResolver.java:127)
	at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:693)
	at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:804)
```

**Comment by jonseah** [comment]
#722 contains the fix

**Comment by jonseah** [comment]
#722 has been released in 2.1.2.RELEASE
Closing issue
